client/src/main/java/io/split/engine/experiments/SplitFetcherImp.java

```diff
@@ -95,36 +116,55 @@ public void run() {
     }
 
     private Set<String> runWithoutExceptionHandling(FetchOptions options) throws InterruptedException, UriTooLongException {
-        SplitChange change = _splitChangeFetcher.fetch(_splitCacheProducer.getChangeNumber(), options);
+        SplitChange change = _splitChangeFetcher.fetch(_splitCacheProducer.getChangeNumber(),
+                _ruleBasedSegmentCacheProducer.getChangeNumber(), options);
         Set<String> segments = new HashSet<>();
 
         if (change == null) {
             throw new IllegalStateException("SplitChange was null");
         }
 
-        if (change.since != _splitCacheProducer.getChangeNumber() || change.till < _splitCacheProducer.getChangeNumber()) {
-            // some other thread may have updated the shared state. exit
+        if (checkExitConditions(change.featureFlags, _splitCacheProducer.getChangeNumber()) ||
+            checkExitConditions(change.ruleBasedSegments, _ruleBasedSegmentCacheProducer.getChangeNumber())) {
             return segments;
         }
 
-        if (change.splits.isEmpty()) {
-            // there are no changes. weird!
-            _splitCacheProducer.setChangeNumber(change.till);
+        if (change.featureFlags.d.isEmpty()) {
+            _splitCacheProducer.setChangeNumber(change.featureFlags.t);
+        }
+
+        if (change.ruleBasedSegments.d.isEmpty()) {
+            _ruleBasedSegmentCacheProducer.setChangeNumber(change.ruleBasedSegments.t);
+        }
+        
+        if (change.featureFlags.d.isEmpty() && change.ruleBasedSegments.d.isEmpty()) {
             return segments;
         }
 
+
         synchronized (_lock) {
             // check state one more time.
-            if (change.since != _splitCacheProducer.getChangeNumber()
-                    || change.till < _splitCacheProducer.getChangeNumber()) {
+            if (checkExitConditions(change.featureFlags, _splitCacheProducer.getChangeNumber()) &&
+                    checkExitConditions(change.ruleBasedSegments, _ruleBasedSegmentCacheProducer.getChangeNumber())) {
```

Comment:
why do you remove `checkReturnConditions`? 

`checkExitConditions` is doing different validations than `checkReturnConditions`